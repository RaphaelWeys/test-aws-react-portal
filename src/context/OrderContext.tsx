import React, { FC, useContext, useState } from 'react';

import { Order } from '../interface/order';

export interface IOrderContext {
  order: Order;
  setOrder: React.Dispatch<React.SetStateAction<Order>>;
}

const OrderContext = React.createContext<IOrderContext | undefined>(undefined);

const OrderProvider: FC = ({ children }) => {
  const [order, setOrder] = useState<Order>({} as Order);

  return <OrderContext.Provider value={{ order, setOrder }}>{children}</OrderContext.Provider>;
};

const useOrder = () => {
  const context = useContext(OrderContext);
  if (context === undefined) {
    throw new Error('useOrder must be used within a OrderContext');
  }
  return context;
};

export { OrderProvider, useOrder };
