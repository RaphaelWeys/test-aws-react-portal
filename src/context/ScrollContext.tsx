import React, { FC, useContext } from 'react';

const ScrollContext = React.createContext<HTMLDivElement | null>(null);

interface IScrollProvider {
  scroll: HTMLDivElement | null;
}

const ScrollProvider: FC<IScrollProvider> = ({ children, scroll }) => {
  return <ScrollContext.Provider value={scroll}>{children}</ScrollContext.Provider>;
};

function useScroll() {
  const context = useContext(ScrollContext);
  if (context === undefined) {
    throw new Error('useScroll must be used within a ScrollProvider');
  }
  return context;
}

export { ScrollProvider, useScroll, ScrollContext };
