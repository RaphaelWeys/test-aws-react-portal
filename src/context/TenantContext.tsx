import { changeAntdTheme } from 'mini-dynamic-antd-theme';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import { useGetTenantInfo } from '../endpoints/tenant/useGetTenantInfo';
import theme, { overrideTheme } from '../style/theme';

const TenantContext = React.createContext(null);

// const domain = 'demo.optimization.energy';
const domain = window.location.hostname;
const regex = /optimization.energy$/;
const shouldGetTenantInfo = domain.match(regex);

const TenantProvider = ({ children }) => {
  const [customTheme, setCustomTheme] = React.useState(theme);
  const [isDomainMA, setIsDomainMA] = React.useState(false);
  const [favicon, setFavicon] = React.useState('');
  const [siteTitle, setSiteTitle] = React.useState('');
  const [tenantName, setTenantName] = React.useState<string | undefined>();
  const [environmentVariables, setEnvironmentVariables] = React.useState(process.env);
  const { mutate: getTenantInfo, isSuccess: getTenantInfoSucceed } = useGetTenantInfo();

  React.useEffect(() => {
    if (!shouldGetTenantInfo) {
      return;
    }

    getTenantInfo(
      { domain },
      {
        onSuccess(data) {
          setEnvironmentVariables((variables) => ({
            ...variables,
            ...data.env,
          }));
          setIsDomainMA(true);
          setSiteTitle(data.siteTitle);
          setFavicon(data.favicon);
          setTenantName(data.tenant);
        },
        onSettled(data) {
          if (data) {
            const { dashboardLogo, loginLogo, ...colors } = data.theme;
            const logo = {
              dashboard: dashboardLogo,
              login: loginLogo,
            };
            setCustomTheme(() => overrideTheme(logo, colors));
            changeAntdTheme(data.theme.baseColor);
          } else {
            changeAntdTheme(theme.colors.baseColor);
          }
        },
      },
    );
  }, [getTenantInfo]);

  if (!getTenantInfoSucceed && shouldGetTenantInfo) {
    return null;
  }

  return (
    <ThemeProvider theme={customTheme}>
      <TenantContext.Provider value={{ env: environmentVariables, isDomainMA, favicon, siteTitle, tenantName }}>
        {children}
      </TenantContext.Provider>
    </ThemeProvider>
  );
};

function useTenant() {
  const context = React.useContext(TenantContext);
  if (context === undefined) {
    throw new Error('useTenant must be used within a TenantProvider');
  }
  return context;
}

export { TenantContext, TenantProvider, useTenant };
