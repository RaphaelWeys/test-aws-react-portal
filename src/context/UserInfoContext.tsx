import React, { FC, useContext, useState } from 'react';

import { useGetUserInfo } from '../endpoints/user/useGetUserInfo';
import { UserInfo } from '../interface/user';

export interface IInfoUserContext {
  userInfo: Partial<UserInfo>;
  setUserInfo: React.Dispatch<React.SetStateAction<Partial<UserInfo>>>;
}

const UserInfoContext = React.createContext<IInfoUserContext | undefined>(undefined);
UserInfoContext.displayName = 'UserInfoContext';

const UserInfoProvider: FC = ({ children }) => {
  const [userInfo, setUserInfo] = useState<Partial<UserInfo>>({});
  const { isFetched, isIdle } = useGetUserInfo(setUserInfo);

  // Prevent to render app, to avoid  unnecessary loading files
  if (!isFetched && !isIdle) {
    return null;
  }

  return <UserInfoContext.Provider value={{ userInfo, setUserInfo }}>{children}</UserInfoContext.Provider>;
};

function useUserInfo() {
  const context = useContext(UserInfoContext);
  if (context === undefined) {
    throw new Error('useUserInfo must be used within a UserInfoProvider');
  }
  return context;
}

export { UserInfoContext,UserInfoProvider, useUserInfo };
