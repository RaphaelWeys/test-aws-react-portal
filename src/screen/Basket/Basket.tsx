import React, { FC } from 'react';
import { useTranslation } from 'react-i18next';
import { Redirect } from 'react-router-dom';
import styled, { css } from 'styled-components';

import marketplaceIcon from '../../assets/picto-marketplace.svg';
import optimizationIcon from '../../assets/picto-optimization.svg';
import Loader from '../../components/Loader';
import { useGetOrder } from '../../endpoints/basket/useGetOrder';
import MainLayout from '../../layout/MainLayout';
import AppBasket from './AppBasket';
import { getQueryParameters } from '../../utils/url';

interface Props {
  className?: string;
}

const Basket: FC<Props> = ({ className }) => {
  const [t] = useTranslation();
  const { orderId, callback } = getQueryParameters();
  const { isLoading, isError, data: order } = useGetOrder(orderId as string);

  if (!callback || order?.paid) {
    return (
      <div data-testid="redirect-component">
        <Redirect to="/" />
      </div>
    );
  }

  if (isLoading) return <Loader />;

  return (
    <MainLayout hasBg={false}>
      <div className={className}>
        {isError ? (
          <h1>{t('basket-order-not-found')}</h1>
        ) : (
          <AppBasket icon={order?.app === 'tender' ? marketplaceIcon : optimizationIcon} />
        )}
      </div>
    </MainLayout>
  );
};

export default styled(Basket)`
  width: 100%;

  h1 {
    font-weight: bold;
    text-align: center;
  }

  ${({ theme: { breakpoints } }) => css`
    @media (min-width: ${breakpoints.lg}) {
      width: 800px;
    }
  `}
`;
