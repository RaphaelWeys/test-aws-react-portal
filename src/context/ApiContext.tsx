import axios, { AxiosInstance } from 'axios';
import React, { FC, useContext, useMemo } from 'react';

import { useUserInfo } from './UserInfoContext';

const ApiContext = React.createContext<AxiosInstance>(axios);

const ApiProvider: FC = ({ children }) => {
  const { userInfo } = useUserInfo();

  const client = useMemo(() => {
    const instance = axios.create({ baseURL: process.env.REACT_APP_BACKEND_PORTAL_URL });

    if (userInfo.token) {
      instance.defaults.headers.common.Authorization = `Bearer ${userInfo.token}`;
    }

    instance.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(error),
    );

    return instance;
  }, [userInfo.token]);

  return <ApiContext.Provider value={client}>{children}</ApiContext.Provider>;
};

function useApi() {
  const context = useContext(ApiContext);
  if (context === undefined) {
    throw new Error('useApi must be used within a ApiProvider');
  }
  return context;
}

export { ApiContext,ApiProvider, useApi };
