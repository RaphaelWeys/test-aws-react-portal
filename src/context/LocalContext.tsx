import { ConfigProvider } from 'antd';
import enUs from 'antd/lib/locale/en_US';
import esES from 'antd/lib/locale/es_ES';
import frFR from 'antd/lib/locale/fr_FR';
import itIt from 'antd/lib/locale/it_IT';
import nlNl from 'antd/lib/locale/nl_NL';
import i18next from 'i18next';
import moment from 'moment';
import 'moment/locale/es';
import 'moment/locale/fr';
import 'moment/locale/it';
import 'moment/locale/nl';
import React, { FC, useState } from 'react';
import { getQueryParameters } from '../utils/url';
import { useUserInfo } from './UserInfoContext';

export type LocalType = 'en' | 'it' | 'fr' | 'es' | 'nl';

export interface ILocalContext {
  local: LocalType;
  setLocal: React.Dispatch<React.SetStateAction<string>>;
}

const LocalContext = React.createContext<ILocalContext | undefined>(undefined);

const LocalProvider: FC = ({ children }) => {
  const { lng } = getQueryParameters();
  const { userInfo } = useUserInfo();
  const { language } = userInfo;
  const [localAnt, setLocalAnt] = useState(undefined);

  React.useEffect(() => {
    if (language === 'it') setLocalAnt(itIt);
    else if (language === 'es') setLocalAnt(esES);
    else if (language === 'fr') setLocalAnt(frFR);
    else if (language === 'nl') setLocalAnt(nlNl);
    else setLocalAnt(enUs);

    moment.locale(language);
    i18next.changeLanguage((lng as string) || language);
  }, [language, lng]);

  if (!localAnt) {
    return null;
  }

  return <ConfigProvider locale={localAnt}>{children}</ConfigProvider>;
};

export { LocalProvider, LocalContext };
