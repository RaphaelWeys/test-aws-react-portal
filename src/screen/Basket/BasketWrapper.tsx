import React, { FC, useContext, useState } from 'react';

import { Order } from '../../interface/order';
import Basket from './Basket';

interface Props {
  className?: string;
}

export interface IOrderContext {
  order: Order;
  setOrder: React.Dispatch<React.SetStateAction<Order>>;
}

const OrderContext = React.createContext<IOrderContext | undefined>(undefined);

export const useOrder = () => {
  const context = useContext(OrderContext);
  if (context === undefined) {
    throw new Error('useOrder must be used within a OrderContext');
  }
  return context;
};

const BasketWrapper: FC<Props> = () => {
  const [order, setOrder] = useState<Order>({} as Order);

  return (
    <OrderContext.Provider value={{ order, setOrder }}>
      <Basket />
    </OrderContext.Provider>
  );
};

export default BasketWrapper;
