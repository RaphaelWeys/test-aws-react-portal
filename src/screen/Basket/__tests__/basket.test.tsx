import 'jest-styled-components';

import { render as rtlRender, waitFor } from '@testing-library/react';
import axios from 'axios';
import React from 'react';
import { Router } from 'react-router-dom';

import { LocalContext } from '../../../context/LocalContext';
import { UserInfoContext } from '../../../context/UserInfoContext';
import history from '../../../router/history';
import Basket from '../Basket';

interface IStateMachineProvider {
  children: React.ReactNode;
}

jest.mock('axios');
jest.mock('little-state-machine', () => ({
  StateMachineProvider: ({ children }: IStateMachineProvider) => <div>{children}</div>,
}));

const data = {
  app: 'tender',
  itemsAmount: 1500,
  couponValue: 500,
  items: [{ quantity: 1, name: 'name-test', description: 'description-test', price: 1500 }],
};

const render = (ui: React.ReactNode) =>
  rtlRender(
    <Router history={history}>
      <UserInfoContext.Provider value={{ userInfo: {}, setUserInfo: jest.fn() }}>
        <LocalContext.Provider value={{ local: 'en', setLocal: jest.fn() }}>{ui}</LocalContext.Provider>
      </UserInfoContext.Provider>
    </Router>,
  );

test('It should use the orderId in the URL to make the request', async () => {
  (axios.request as jest.Mock).mockResolvedValue({ data });
  const orderId = 123456789;
  const fakeCallback = 'https://fake-address.com';
  const { location } = window;
  delete window.location;
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  window.location = { assign: jest.fn() };
  window.location.search = `orderId=${orderId}&callback=${fakeCallback}`;
  const { queryByRole } = render(<Basket />);

  expect(queryByRole('img')).toBeInTheDocument();

  await waitFor(() =>
    expect(axios.request).toHaveBeenCalledWith({
      headers: { 'Content-Type': 'application/json' },
      method: 'get',
      url: `/order/${orderId}`,
    }),
  );

  window.location = location;
});

test('It should match snapshot', async () => {
  (axios.request as jest.Mock).mockResolvedValue({ data });
  const { container } = render(<Basket />);

  await waitFor(() => expect(axios.request).toHaveBeenCalled());

  expect(container).toMatchSnapshot();
});

test('It should show an appropriate message if there is an error or empty order', async () => {
  (axios.request as jest.Mock).mockRejectedValueOnce({
    response: {
      status: 400,
    },
  });
  const fakeCallback = 'https://fake-address.com';
  const { location } = window;
  delete window.location;
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  window.location = { assign: jest.fn() };
  window.location.search = `callback=${fakeCallback}`;

  const { getByText } = render(<Basket />);

  await waitFor(() => expect(axios.request).toHaveBeenCalled());

  expect(getByText('basket-order-not-found')).toBeInTheDocument();
  window.location = location;
});

test('It should redirect to home page if no callback are in URL', async () => {
  const { getByTestId } = render(<Basket />);

  await waitFor(() => expect(getByTestId(/redirect-component/i)).toBeInTheDocument());
});

test('It should redirect to home page if order is already paid', async () => {
  (axios.request as jest.Mock).mockResolvedValue({ data: { ...data, paid: true } });
  const fakeCallback = 'https://fake-address.com';
  const { location } = window;
  delete window.location;
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  window.location = { assign: jest.fn() };
  window.location.search = `callback=${fakeCallback}`;

  const { getByTestId } = render(<Basket />);

  await waitFor(() => expect(getByTestId(/redirect-component/i)).toBeInTheDocument());

  window.location = location;
});
