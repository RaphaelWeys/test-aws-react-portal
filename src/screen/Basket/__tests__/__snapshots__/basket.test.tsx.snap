// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should match snapshot 1`] = `
<div>
  <div
    data-testid="redirect-component"
  />
</div>
`;
